lucille-core/src/main/java/com/kmwllc/lucille/stage/FetchUri.java

```diff
@@ -0,0 +1,130 @@
+ *  e.g. url --> url_(size_suffix) where source name is url
+ * status_suffix (String, Optional) : suffix to be appended to end of source field name for the status code of the fetch request
+ * error_suffix (String, Optional) : suffix to be appended to end of source field name for any errors in process
+ * max_size (Integer, Optional) : max size, in bytes, of data to be read from fetch
+ */
+public class FetchUri extends Stage {
+
+  private final String source;
+  private final String dest;
+  private final String statusSuffix;
+  private final String sizeSuffix;
+
+  private final String errorSuffix;
+  private final int maxDownloadSize;
+  private CloseableHttpClient client;
+
+  public FetchUri(Config config) {
+    super(config, new StageSpec().withRequiredProperties("source", "dest")
+        .withOptionalProperties("size_suffix", "status_suffix", "max_size", "error_suffix"));
+    this.source = config.getString("source");
+    this.dest = config.getString("dest");
+    this.statusSuffix = ConfigUtils.getOrDefault(config, "status_suffix", "status_code");
+    this.sizeSuffix = ConfigUtils.getOrDefault(config, "size_suffix", "size");
+    this.errorSuffix = ConfigUtils.getOrDefault(config, "error_suffix", "error");
+    this.maxDownloadSize = ConfigUtils.getOrDefault(config, "max_size", Integer.MAX_VALUE);
+  }
+
+  // Method exists for testing with mockito mocks
+  void setClient(CloseableHttpClient client) {
+    this.client = client;
+  }
+
+  @Override
+  public void start() throws StageException {
+    client = HttpClients.createDefault();
+  }
+
+  @Override
+  public Iterator<Document> processDocument(Document doc) throws StageException {
+    if (!doc.has(source) || doc.getString(source).isEmpty()) {
+      return null;
+    }
+    String url = doc.getString(source);
+
+    // Following checks for valid URL
+    try {
+      new URL(url).toURI();
+    } catch (URISyntaxException e) {
+      setErrorField(doc, e);
+      return null;
+    } catch (MalformedURLException e) {
+      setErrorField(doc, e);
+      return null;
+    }
+
+    HttpGet httpGet = new HttpGet(url);
+    HttpEntity ent = null;
+    try (CloseableHttpResponse httpResponse = client.execute(httpGet)) {
+      int statusCode = httpResponse.getStatusLine().getStatusCode();
+      ent = httpResponse.getEntity();
+      InputStream content = ent.getContent();
+      BoundedInputStream boundedContentStream = new BoundedInputStream(content, maxDownloadSize);
+      byte[] bytes = IOUtils.toByteArray(boundedContentStream);
+      long contentSize = bytes.length;
+
+      doc.setField(dest, bytes);
+      doc.setField(source + "_" + statusSuffix, statusCode);
+      doc.setField(source + "_" + sizeSuffix, contentSize);
+    } catch (ClientProtocolException e) {
+      setErrorField(doc, e);
+    } catch (IOException e) {
+      setErrorField(doc, e);
+    } finally {
+      try {
+        EntityUtils.consume(ent);
```

Comment:
If ent is null here will EntityUtils handle it okay? Just checking if we need our own null check here.

Reply:
Looking at the code for .consume, it checks whether it's null. It doesn't do anything if it's null

Reply:
`public static void consume(HttpEntity entity) throws IOException {
    if (entity != null) {
      if (entity.isStreaming()) {
        InputStream inStream = entity.getContent();
        if (inStream != null) {
          inStream.close();
        }
      }

    }
  }`

Reply:
I'm wondering if we need EntityUtils.consume(ent) at all now... all it does is close the content InputStream.
We're creating a BoundedInputStream from ent.getContent() in a try-with-resources.
It looks like BoundedInputStream.close() propogates the close() to the underling InputStream when isPropagateClose() is true. Looking at code for BoundedInputStream, true is the default. Thoughts?
https://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/input/BoundedInputStream.html#close()

Reply:
I looked at the source code for both .close() and .consume(which uses .close()) and I think that we can remove the .consume. I also was reading this https://stackoverflow.com/questions/65298317/how-does-consuming-till-the-end-of-stream-using-entityutils-consumehttpentity and came to the same conclusion. With .consume, if the .getContent() from the method creates an IOException, the stream wouldn't close. Would this cause us to favor .close() instead?

Reply:
Yeah I agree. The EntityUtils.consume can be removed and also the finally block.

Ultimately, we want all of the closeable stuff that we've opened to be closed cleanly and consistently which I think now that we have all of the Closable things in try-with-resources that is achieved without needing EntityUtils.